### ISSUE TITLE
Take a look at CFU-playground from Google?

### ISSUE THREAD

**Issue by mithro** [cbalint13/OLIMP#1]
Hi!

I thought you might find the CFU-Playground from Google interesting. It is a tool which is designed to speed up the ["OODA loop" (Observe Orient Decide Act)](https://en.wikipedia.org/wiki/OODA_loop) in CPU opcode design. It seems like OLIMP might be interesting to try inside the project. Links are;
 * http://cfu-playground.rtfd.io/
 * https://github.com/google/cfu-playground

Hope it is interesting to you!

Tim '@mithro' Ansell

**Comment by cbalint13** [comment]
Hi Tim !

First, thank you for your review-time and insights !

> I thought you might find the CFU-Playground from Google interesting. It is a tool which is designed to speed up the ["OODA loop" (Observe Orient Decide Act)](https://en.wikipedia.org/wiki/OODA_loop) in CPU opcode design. It seems like OLIMP might be interesting to try inside the project. Links are;

Yes, was aware of these efforts, they are **very** interesting to me !

> * http://cfu-playground.rtfd.io/
> * https://github.com/google/cfu-playground

**BUT,**

OODA effort (the ML related part) is similar as OLIMP goal but in my view lacks efforts toward a more universal/fundamental approach (i perceived no signs/directions there to tackle it properly).

But we can change it, will try to discuss OODA vs OLIMP goals with the community there and see if we can have common benefits (would even join their efforts too).

 I take a  first step, right here, inviting Cc: @tcal-x @alanvgreen (thanks them in advance)

-----------

I detail two fundamental things that OLIMP relays, contrasting OODA's ML part:

1. How any arbitrary designed HW block (as SIMD ext. or larger PE) can fit into any arbitrary computing graph ?  
    * The logarithmic **key** solution (as basic building block) so far is UNIT: https://arxiv.org/abs/2101.08458 (for tensorization)
    * Any HW operator/module (let it arbitrary) can fit "guaranteed" optimally into any arbitrary ML compute graph.

2. Also, need a flexible enough ml-software framework to fit this into:
    * I found TVM (Apache TVM) expressive enough also having ML driven search space exploration (a very powerful one). 
    * TVM expressive IR engine is capable of optimizing itself, contrasting with tensorflow/pytorch "hand optimized" fixed SIMD approach.
    * TVM 's also made steps supporting "verilated" SIMD/HW backend right now, it offers this "flexible HW" in their devel goals. 

OLIMP's ultimate goal is:  Standalone HW (HDL side)  & ML (compute graph side) joint co-optimization by ML itself.

**Comment by alanvgreen** [comment]
Hi Balint,

Wow. So many new ideas here! Thank you for sharing them so freely! I'd like to make sure I understand what OLIMP does.

1. TfLite Micro is an interpreter that executes ML operators. The implementation of these ML operators can be hand optimized to take advantage of hardware acceleration, which is the approach the CFU-Playground takes.

2. In contrast, TVM compiles the model directly to code, with machine-optimized instruction sequences that take into account the parameters to each ML operator and the hardware it is targeted for.

3. OLIMP takes TVM and some interesting hardware accelerator blocks and combines them for end-to-end acceleration that does not require additional hand-optimization.

Is all this right? I think I might be missing something in considering how ML optimization works.

Alan

**Comment by cbalint13** [comment]
> Hi Balint,

Hi Alan,

> 
> Wow. So many new ideas here! Thank you for sharing them so freely! I'd like to make sure I understand what OLIMP does.


 I also thank you much for your time !

 
> 1. TfLite Micro is an interpreter that executes ML operators. The implementation of these ML operators can be hand optimized to take advantage of hardware acceleration, which is the approach the CFU-Playground takes.

Yes I noticed that, also familiar with TF/TFLite way of handling / passing through the graph.


> 2. In contrast, TVM compiles the model directly to code, with machine-optimized instruction sequences that take into account the parameters to each ML operator and the hardware it is targeted for.

Yes but No. 


* TVM compiles first to IR+TIR (elegant and powerful ones) then jointly optimize IR->TIR->SIMD all-in-one as "opportunities". 
* TVM doesn't know much (zero in latest [TVM2.0](https://arxiv.org/pdf/2006.06762.pdf)) about hardware it only knows some generic receipt for dense, conv2d, etc.
* SIMD opportunities are discovered by space search (using a ML/XGboost costs model) on many measured "kernel" variants.

It is able to universally catch&tune fine grained HW details, **without telling much about the hardware itself**.

> 3. OLIMP takes TVM and some interesting hardware accelerator blocks and combines them for end-to-end acceleration that does not require additional hand-optimization.


Exactly, _no_ hand optimization, the picture is:

 1. OLIMP emit a possible HW/HDL unit, using user input constrains like: amount of DSP, memory, bandwidth, layout-flavour
 2. Using approach like  [UNIT](http://www.seas.ucla.edu/~jianw/unit-cgo2021.pdf) the HW unit is fitted right in to the TIR(IR) most inner loops of operators (optimally guaranteed).
 3. TVM further optimize jointly the whole NN graph IR+TIR and may shed even better opportunities related to whole graph.
 4. Final emitted code (can be C or llvm/binary) contains SIMD calls to HW/unit via extended ISA or register-calls to some mem mapped block (indirect).

See a concrete sample, dissecting OLIMP:

  * This is a generated [TIR](https://github.com/cbalint13/OLIMP/tree/main/demo/icebreaker) with a simple OLIMP VEC-8U8-16I8-2S32 [notice **tir.call_extern("O_VEC_MACC" )**] for a **dense** op.
  * For now HW tensorization is encoded [manually](https://github.com/cbalint13/OLIMP/blob/main/srcs/sims/olimp-vec-gemm.py#L290) into a **dense-op** after little [loop reordering](https://github.com/cbalint13/OLIMP/blob/main/srcs/sims/olimp-vec-gemm.py#L281) taking place.
  * Final code will contain call to ISA (extension) or call to some registers calling a PE or [whatever user function](https://github.com/cbalint13/OLIMP/blob/main/srcs/sims/olimp-vec-gemm.py#L32)

   The "manual" approach will be replaced having [UNIT](http://www.seas.ucla.edu/~jianw/unit-cgo2021.pdf) approach.

There is a major refactor [WiP](https://github.com/apache/tvm/issues/7527) on TVM side to include mentioned [TVM2.0](https://arxiv.org/pdf/2006.06762.pdf)  then [UNIT](http://www.seas.ucla.edu/~jianw/unit-cgo2021.pdf) then OLIMP can make it's way nicely upon.


> Is all this right? I think I might be missing something in considering how ML optimization works.

Well, there will be always optimization opportunities even beyond of some efforts described here.
But having [UNIT](http://www.seas.ucla.edu/~jianw/unit-cgo2021.pdf) in first is a neat way to fit a "given" unit into a larger operator so things can be automated much.

-----

I spent lot of time on polyhedral approaches and wasn't able to do a "UNIT" like simple algo, searched into academia too, but it just happened recently (incidentally relaying right on TVM internals).

> Alan

~Cristian

**Comment by alanvgreen** [comment]
Hi Cristian,

Super interesting. Did you have ideas about integrating this with CFU Playground?

I'll take some time to think about ways these projects can help each other.

Alan

**Comment by tcal-x** [comment]
Hi @cbalint13 , I finally got a bit of time to dig in to OLIMP and the demo SoC, and I think I'm starting to get it.  Please correct me if I'm wrong:

The coef and data memory blocks are dedicated to the vector/UNIT hw block; they are connected to the main 32b memory bus and can be written by the CPU; they are in turn read at wider widths (128b and 64b) by the UNIT block.

CPU instructions coordinate UNIT execution through the [PCPI (coprocessor interface) ](https://github.com/cliffordwolf/picorv32#pico-co-processor-interface-pcpi) by providing the addresses for both coef and data memories, and also the control that "this is a valid coprocessor instruction, so update your accumulator accordingly" (whether it  is an initial-zero MAC or a running MAC).

This is a bit different than our handmade examples, where the address generation and sequencing is done by FSMs *inside* our coprocessor/accelerator.    The OLIMP approach is a nice design point in that the custom synthesized hardware is more straightforward, and the addressing all remains in software.

I expect you're considering having multiple UNITs in one system, for example, one optimized to support CONV2D and another to support DEPTHWISE_CONV, and perhaps even specialized depending on whether the filter is  3x3 vs 1x1, for example?   In such a case, I would expect the coef and data memories to be shared by the different UNITs.

**Comment by cbalint13** [comment]
> Hi Cristian,
> 
> Super interesting. Did you have ideas about integrating this with CFU Playground?

* Huh, good question.

   The idea to have an algo capable to fit ("divide") a HW unit (expressed by it's small compute graph) into a larger compute graph (any operator CNN, DENSE or other more complicated fused form) is big a step toward automated tensorization.
   UNIT comes as a compiler pass that do this, but its doable at this moment only on TVM https://github.com/were/UNIT

 Unsure how to tackle it exactly into TF's MLIR schema and the "grappler" but is not imposibile (as far as i checked last time).

> I'll take some time to think about ways these projects can help each other.
> 
> Alan

**Comment by cbalint13** [comment]
> Hi @cbalint13 , I finally got a bit of time to dig in to OLIMP and the demo SoC, and I think I'm starting to get it. Please correct me if I'm wrong:

That SoC example is atempt to fit a small riscv32 + olimp inside less then 5k LUT (Lattice ICE40), a challenging attempt i would say.


> The coef and data memory blocks are dedicated to the vector/UNIT hw block; they are connected to the main 32b memory bus and can be written by the CPU; they are in turn read at wider widths (128b and 64b) by the UNIT block.

  * Yes, 64b and 128b bus are demanded by DSP-MAC in that particular OLIMP layout, also the write-backs have to be multiple 64/32b pipes. Unfortunate is that this ICE40 memory blocks are bit limited in amount and I am also dancing at the very edge of my 20Mhz target closure.

  * But is doable in this way, goal is to attract attention that this is possible in such small amount of LUT.



> CPU instructions coordinate UNIT execution through the [PCPI (coprocessor interface) ](https://github.com/cliffordwolf/picorv32#pico-co-processor-interface-pcpi) by providing the addresses for both coef and data memories, and also the control that "this is a valid coprocessor instruction, so update your accumulator accordingly" (whether it is an initial-zero MAC or a running MAC).

* Exactly, PCPI of picorv32 is very handy to extend the rv32 ISA with "whatever".
* I will pass (already) via CPU registers (that PCPI handles) the HW pointers to very coef & data.
* Results (write-backs) will be done by that same OLIMP unit but at different memory (BRAM) partitions (having less bus-wide).
* Anyway memory partitioned physically in such strange BRAM groups can be made to look the "same" to the very CPU.
* Software calls can be encapsulated as _asm_( .word 0x) inline, TVM handle this well within it's very LLVM backend (once tensorized).

This is only a small study case on the small ICE40, i plan more advanced show cases on: https://github.com/cbalint13/e-verest of FPGA vendor, and can be extendend to ASIC design if one afford and see scopes in it.

e-Verest came in mind due that some people (including me) may have "alergy" to cables and wires, they might want just a simple re-usable stick.

> > This is a bit different than our handmade examples, where the address generation and sequencing is done by FSMs _inside_ our coprocessor/accelerator. The OLIMP approach is a nice design point in that the custom synthesized hardware is more straightforward, and the addressing all remains in software.

Yes, this is a the difference.

The nice part will come will the ability (already) of TVM to emit properly tensorized code out of the box end-to-end. Wherever there are operations that cannot leverage/catch the extended ISA-OLIMP, for some reasons UNIT pass fail to tensorize,  the normal rv32im ISA (expensive and slow) will kick-in since rv32im have sufficient muls&divs&bitshifts to cover any other NN layers that cannot be covered otherwise (done by actual LLVM engine inside TVM).

> I expect you're considering having multiple UNITs in one system, for example, one optimized to support CONV2D and another to support DEPTHWISE_CONV, and perhaps even specialized depending on whether the filter is 3x3 vs 1x1, for example?

* UNIT is just a "pass" an [algorithm that tensorize](https://github.com/were/UNIT), a plugin inside TVM, it embeds/mangle automatically a OLIMP operator into the computation graph. It have the ability to incorporate into any kind of CONV2 / 3D or other more complex forms. The variants of CONV2D doesn't matter as long it is able to replace efficiently with a tensorized version.
* Yes, i expect that for some conv2d variants tensorization will not go as well  (in performance terms) as for other but there is also TVM ability to explore variations by it's existent "autotune" function.
* But it will be possible to have multiple OLIMP HW  operators (as multiple extended ISA ops) and letting UNIT compiler pass to find ways to tensorize them optimally into the graph of NN. Not exprimented yet so far, but its a goal.

> >  In such a case, I would expect the coef and data memories to be shared by the different UNITs.

Yes, in this case BRAM and memory organization grows in complexity, bandwith and instant access is crucial&mandatory.

* My plan is simple, in two steps: just group enough BRAM around each very OLIMP unit, exactly how the OLIMP unit demands the access (having hungry DSPs inside), but at same time make all these BRAM to look **contiguous** for the very main CPU, thus software side can see it all (via 32bit bus indeed). 

I consider a more generalized framework for FPGA primitives (LUT+BRAM+DSP), that can capture and manipulate via expressible IR beside a simple compiler tool, capable to emit vendor independet verilog code:
     - https://github.com/vegaluisjose/reticle
     - https://www.cs.cornell.edu/~asampson/media/papers/reticle-pldi2021.pdf

* There will be more fun when will attempt to do more complex blocks in OLIMP like a systolic array or a more specialized/fixed CONV units. They (booth) still can leverage the UNIT algorithm (pass) and make it's way to magic tensorization but i expect cases where UNIT may fail to find "tensorizated" solution to include such larger blocks.
